File: lib/Controller/Controller.cpp
#include "Controller.h"

Controller::Controller(float Kp, float Ki, float Kd, float Kf) {
  this->Kp = Kp;
  this->Ki = Ki;
  this->Kd = Kd;
  this->Kf = Kf;
  Setpoint = Dt = Error = Integral = LastTime = LastError = 0;        
}

void Controller::setPIDF(float Kp, float Ki, float Kd, float Kf) {
  this->Kp = Kp;
  this->Ki = Ki;
  this->Kd = Kd;
  this->Kf = Kf;
}

void Controller::reset(){
  this->LastTime = 0;
  this->Integral = 0;
  this->LastError = 0;
  this->LastTime = 0;
}

float Controller::Calculate(float setpoint, float measure){
  Setpoint = setpoint;
  return Calculate(setpoint - measure);
}

float Controller::Calculate(float error){
  unsigned long CurrentTime;
  // Serial.println(error);
  CurrentTime = millis();
  Dt = CurrentTime / 1000.0 - LastTime / 1000.0;
  // Serial.println(Dt);
  LastTime = CurrentTime;
  Error = error;
  Integral += Error * Dt;
  Integral = constrain(Integral, -100, 100);
 if (error == 0) Integral = 0;
  float Derivative = (Error - LastError) / Dt;
  LastError = Error;
  return Kp * Error + Ki * Integral + Kd * Derivative + Kf * SigNum(error);
}

byte Controller::SigNum(float number) {
    return (byte) (number == 0 ? 0 : (number < 0 ? -1 : 1));
}

File: robot_ws/src/robot_core/src/controller.py
from src.utilize import *
import time

class Controller:
    # pidf Setting
    kp = 0
    ki = 0
    kd = 0
    kf = 0
    baseSpeed = 0
    
    Setpoint = 0
    Error = 0
    ErrorTolerance = 0
    LastError = 0
    Integral = 0 
    Dt = 0
    CurrentTime = time.time()
    LastTime = CurrentTime
    
        
    def __init__(self, kp = 0, ki = 0, kd = 0, kf = 0, baseSpeed = 0, errorTolerance = 0):
        self.ConfigPIDF(kp, ki, kd, kf, baseSpeed)
        self.ErrorTolerance = abs(errorTolerance)
        self.LastTime = time.time()
        
    def ConfigPIDF(self, kp = 0, ki = 0, kd = 0, kf = 0, baseSpeed = 0):
        self.kp = kp
        self.ki = ki
        self.kd = kd
        self.kf = kf
        self.baseSpeed = baseSpeed
    
    def Calculate(self, setpoint, measurement):
        return self.Calculate(setpoint - measurement)
    
    def Calculate(self, error):
        self.CurrentTime = time.time()
        self.Dt = self.CurrentTime - self.LastTime
        self.LastTime  = self.CurrentTime
        Is_Error_In_Tolerance = AtTargetRange(error, 0, self.ErrorTolerance)
        if Is_Error_In_Tolerance :
            return 0
        self.Error = error
        self.Integral += self.Error * self.Dt
        self.Integral = clip(self.Integral, -1, 1)
        Derivative = (self.Error - self.LastError) / self.Dt if self.Dt != 0 else 0
        self.LastError = self.Error
        BaseSpeed = self.baseSpeed * sig_num(self.Error)
        return (self.Error * self.kp) + (self.Integral * self.ki) + (Derivative * self.kd) + (self.Setpoint * self.kf) + BaseSpeed
    
    def ResetVariable(self):
        self.LastError = 0
        self.Integral = 0 
        self.CurrentTime = time.time()
        self.LastTime = self.CurrentTime
        
    def Reset(self, kp = None, ki = None, kd = None, kf = None, baseSpeed = None):
        # pidf Setting
        kp = self.kp if kp == None else kp
        ki = self.ki if ki == None else ki
        kd = self.kd if kd == None else kd
        kf = self.kf if kf == None else kf
        baseSpeed = self.baseSpeed if baseSpeed == None else baseSpeed
        self.ConfigPIDF(kp, ki, kd, kf, baseSpeed)
            
        # pid variable
        self.Setpoint = 0
        self.Error = 0
        self.LastError = 0
        self.Integral = 0 
        self.Dt = 0
        self.CurrentTime = time.time()
        self.LastTime = self.CurrentTime
